tensorboard/data/provider.py

```diff
@@ -0,0 +1,223 @@
+        series will only be included in the result if its run and tag
+        both pass this filter. If `None`, all time series will be
+        included.
+      step_filter: Optional `StepFilter` value. If `None`, the entire
+        range of steps may be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a list of
+      `ScalarDatum` values sorted by step.
+    """
+    pass
+
+  def list_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def read_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def list_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+  def read_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+
+_ScalarMetadata = collections.namedtuple(
+    "ScalarMetadata",
+    ("max_step", "corresponding_wall_time", "summary_metadata"),
+)
+class ScalarMetadata(_ScalarMetadata):
+  """Metadata about a scalar time series for a particular run and tag.
+
+  Attributes:
+    max_step: The largest step for this scalar time series; an integer.
+    corresponding_wall_time: The wall time of the datum with the largest
+      step in this time series, as `float` seconds since epoch.
+    summary_metadata: A `summary_pb2.SummaryMetadata` value.
+  """
+  pass
+
+
+_ScalarDatum = collections.namedtuple(
+    "ScalarDatum", ("step", "wall_time", "value")
+)
+class ScalarDatum(_ScalarDatum):
+  """A single datum in a scalar time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This
+      is a unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as
+      `float` seconds since epoch.
+    value: The scalar value for this datum; a `float`.
+  """
+  pass
+
+
+class RunTagFilter(object):
+  """Filters data by run and tag names."""
+
+  def __init__(self, runs=None, tags=None):
+    """Construct a `RunTagFilter`.
+
+    A time series passes this filter if both its run *and* its tag are
+    included in the corresponding whitelists.
+
+    Order and multiplicity are ignored; `runs` and `tags` are treated as
+    sets.
+
+    Args:
+      runs: Collection of run names, as strings, or `None` to admit all
+        runs.
+      tags: Collection of tag names, as strings, or `None` to admit all
+        tags.
+    """
+    self._runs = None if runs is None else frozenset(runs)
+    self._tags = None if tags is None else frozenset(tags)
+
+  @property
+  def runs(self):
+    return self._runs
+
+  @property
+  def tags(self):
+    return self._tags
+
+  def test(self, run, tag):
+    if self._runs is not None and run not in self._runs:
+      return False
+    if self._tags is not None and tag not in self._tags:
+      return False
+    return True
+
+
+class StepFilter(object):
```

Comment:
Design question, but would we ever want a filter on wall_time or relative time as well?  This could come up if imagine implementing efficient zooming, and supported all 3 x-axis options (not just step).

Doesn't have to be addressed here but thought I'd mention it since maybe we want to leave room for this in the API somewhere, e.g. by making this something more like TimeAxisFilter?

Reply:
Yes. I was actually planning to rename this to `TimeAxisFilter`
specifically, so I’m glad that you agree. I’ll post more details later.